corehq/apps/hqwebapp/oauth_scopes.py

```diff
@@ -0,0 +1,86 @@
+import json
+import os
+from collections import defaultdict
+
+from oauth2_provider.scopes import BaseScopes
+
+from corehq.apps.domain.utils import get_domain_from_url
+from corehq.motech.fhir.const import FHIR_VERSION_4_0_1
+from corehq.motech.fhir.models import FHIRResourceType, get_schema_dir
+
+DEFAULT_SMART_SCOPES = {
+    'launch/patient': 'Access your unique patient ID',
+    'offline_access': 'Maintain access to your data when you are offline',
+    'online_access': 'Maintain access to your data when you are online',
+}
+
+HQ_DEFINED_SCOPES = {
+    'access_apis': 'Access API data on all your CommCare projects',
+}
+
+
+class FhirScopeBuilder():
+    CLIENT_TYPES = [('patient', 'your patients\''), ('user', 'your')]
+    ACCESS_PERMISSIONS = [('read', 'Read'), ('write', 'Write'), ('*', 'Read and write')]
+
+    def __init__(self, fhir_version=FHIR_VERSION_4_0_1):
+        self.fhir_version = fhir_version
+
+        self.fhir_scopes = {**DEFAULT_SMART_SCOPES}
+        self.fhir_scopes_by_resource_name = defaultdict(list)
+
+        self._build_scope_messages()
+
+    def _build_scope_messages(self):
+        # Build all possible SMART-on-FHIR clinical and launch scopes
+        # http://hl7.org/fhir/smart-app-launch/scopes-and-launch-context/index.html#clinical-scope-syntax
+        # clinical-scope ::= ( 'patient' | 'user' ) '/' ( fhir-resource | '*' ) '.' ( 'read' | 'write' | '*' )
+        for client_type, client_message in self.CLIENT_TYPES:
+            for resource_name, description in self._get_fhir_resources() + [('*', 'all')]:
+                for access_type, access_type_message in self.ACCESS_PERMISSIONS:
+                    message = f"{access_type_message} {client_message} {resource_name} data: {description}"
```

Comment:
this is going to make adding translations to this tricky (not a big deal, just observing)

Reply:
Yeah, I wasn't sure how we would do translations here, since all the text currently comes from a JSON file. I'd also be happy to just drop the `client_message` part of this as it is a bit verbose and confusing.